### PR TITLE
Trevor encoder fixup

### PR DESCRIPTION
**Pull request by TrevorDecker**
This fixes the problem with incorrect encoder messages

The problem was with how we were handling the reading of messages from a serial bus.  

The problem case was as follows:

suppose that  a sensor sends the following bytes were each leading letter represents a different message  At the beginning the sensor bus is:  A_header, A_body_1, A_body_2, A_body_3, A_body_4, Footer, B_header, B_body_1,B_body_2,B_body_3,B_body_4, B_Footer, C_header, C_body_1,C_body_2,C_body_3,C_body_4, C_Footer

now suppose that serial node would read 7 bytes into buf.  Then peel is called and the first 6 bytes are used leaving buf[0] = B_header. At this point the read thread sleeps.  When it reawakens  assume that the serial bus has gotten more data causing the available bus to be: C_body_1,C_body_2,C_body_3,C_body_4, C_Footer

These 5 bytes are read and the buffer is now: B_header,C_body_1,C_body_2,C_body_3,C_body_4, C_Footer   which will appear to be a valid message for message of type B. However it will have data from message type C.  

This problem was fixed by dropping the bytes from previous reads every time that the serial read thread resumes.  This approach does have the downside of dropping messages but stops messages from appearing as the wrong type.

### CODE REVIEW

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package com.roboclub.robobuggy.ui;
+import java.util.Random;
+
+import javax.swing.JPanel;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'javax.swing.JPanel'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l4)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.roboclub.robobuggy.messages.EncoderMeasurement;
+import com.roboclub.robobuggy.messages.SteeringMeasurement;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.messages.SteeringMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l7)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package com.roboclub.robobuggy.ui;
+import java.util.Random;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Random'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l2)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import com.roboclub.robobuggy.messages.EncoderMeasurement;
+import com.roboclub.robobuggy.messages.SteeringMeasurement;
+import com.roboclub.robobuggy.ros.Message;
+import com.roboclub.robobuggy.ros.MessageListener;
+import com.roboclub.robobuggy.ros.SensorChannel;
+import com.roboclub.robobuggy.ros.Subscriber;
+
+import org.jfree.chart.*;
+import org.jfree.chart.plot.PlotOrientation;
+import org.jfree.data.xy.XYDataset;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'org.jfree.data.xy.XYDataset'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l15)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java [hunk 5]
```diff
@@ -42,6 +42,7 @@ public double getLongitude(){
 	private int frameWidth;
 	private int frameHeight;
 	private Subscriber gpsSub;
+	private Subscriber encSub;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Avoid unused private fields such as 'encSub'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java [hunk 6]
```diff
@@ -34,7 +34,7 @@
 	private final Queue<String> _logQueue;
 	private final ArrayList<Subscriber> subscribers;
 
-	private static final Queue<String> startLoggingThread(PrintStream stream) {
+	private static final Queue<String> startLoggingThread(final PrintStream stream) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid really long methods.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642463/issues/source?bid=2938665&fileBranchId=3058909#l37)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 7]
```diff
@@ -49,8 +51,8 @@
 	public static  String COM_PORT_IMU = "COM6";
 	public static  String COM_PORT_GPS_INTEGRATED = "COM7";
 	public static  String COM_PORT_GPS_STANDALONE = "COM8";
-	public static  String COM_PORT_ENCODER = "COM3";
-	public static  String COM_PORT_DRIVE_CONTROL = "COM9";
+	public static  String COM_PORT_ENCODER = "/dev/tty.usbmodem1411";
+	public static  String COM_PORT_DRIVE_CONTROL = "/dev/tty.usbmodem1411";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l55)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java [hunk 8]
```diff
@@ -49,8 +51,8 @@
 	public static  String COM_PORT_IMU = "COM6";
 	public static  String COM_PORT_GPS_INTEGRATED = "COM7";
 	public static  String COM_PORT_GPS_STANDALONE = "COM8";
-	public static  String COM_PORT_ENCODER = "COM3";
-	public static  String COM_PORT_DRIVE_CONTROL = "COM9";
+	public static  String COM_PORT_ENCODER = "/dev/tty.usbmodem1411";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l54)

**Comment by zsd-planekid** [comment]
Zach ok's

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 1 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'javax.swing.JPanel'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l4)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 2 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.messages.SteeringMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l7)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Random'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l2)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'org.jfree.data.xy.XYDataset'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l15)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java (hunk 5 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Avoid unused private fields such as 'encSub'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java (hunk 6 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid really long methods.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642463/issues/source?bid=2938665&fileBranchId=3058909#l37)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 7 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l55)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 8 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l54)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 1 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'javax.swing.JPanel'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l4)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 2 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.messages.SteeringMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l7)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Random'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l2)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/EncoderGraph.java]: Codacy Issue found: [Avoid unused imports such as 'org.jfree.data.xy.XYDataset'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642522/issues/source?bid=2938665&fileBranchId=3058909#l15)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java (hunk 5 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Avoid unused private fields such as 'encSub'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642523/issues/source?bid=2938665&fileBranchId=3058909#l45)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java (hunk 6 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/logging/SensorLogger.java]: Codacy Issue found: [Avoid really long methods.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642463/issues/source?bid=2938665&fileBranchId=3058909#l37)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 7 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l55)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java (hunk 8 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/main/config.java]: Codacy Issue found: [The field name indicates a constant but its modifiers do not](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/1743642470/issues/source?bid=2938665&fileBranchId=3058909#l54)

**Pull request merged**